### PR TITLE
init: Fixes for file descriptor accounting

### PR DESCRIPTION
**Pull request by tryphe**
**Bug 1:** If the daemon does not have enough file descriptors (we'll call them FDs), it is not asserted. 
The daemon only aborts if `FD count < MIN_CORE_FILEDESCRIPTORS(150)`.

**Bug 2:** (Unsure how to reproduce undefined behavior) If the system ulimit setting is a low number, the daemon does not allocate the right amount of FDs during init via `RaiseFileDescriptorLimit`. But it's close.

**Steps to produce Bug 1**: Run `ulimit -n 150`, then start `bitcoind` in the same shell.

**Result:** bitcoind runs normally and displays an erroneous warning:
~~~
Warning: Reducing -maxconnections from 125 to -8, because of system limitations.
...
Using at most -8 automatic connections (150 file descriptors available)
~~~

**Expected:** bitcoind should not start, as it needs roughly 163~172~(?) FDs.

-`MIN_CORE_FILEDESCRIPTORS = 150`
-`MAX_ADDNODE_CONNECTIONS = 8` (new in assertion)
~-`MAX_OUTBOUND_CONNECTIONS = 8`~
~-`CConnman::Options:nMaxFeeler = 1`~(Edit: these exist within the bounds of -maxconnections, if it's high enough)
-Number of `-bind` interfaces, default = 1 (new in allocation)
-Number of `-rpcthreads`, default = 4 (new in allocation and assertion)

150 + 8 + 1 + 4 = 163 minimum
+125 maxconnections = 288 speculative maximum

**Coverage with this patch + `ulimit -n 150`:**
~~~
Bitcoin Core version v0.18.99.0-50ccaa56f -dirty (release build)
Error: Not enough file descriptors available. 150 available, 163 required.
~~~
*daemon closes*

**Coverage with this patch + `ulimit -n 163`:**
~~~
Bitcoin Core version v0.18.99.0-50ccaa56f (release build)
There are 163 file descriptors available, 163 required, 163 reserved, and 288 requested.
Warning: Reducing -maxconnections from 125 to 0, because of file descriptor limitations.
...
Using at most 0 automatic connections (163 file descriptors available)
~~~

**Coverage with this patch + `ulimit -n 203`:**
~~~
Bitcoin Core version v0.18.99.0-50ccaa56f (release build)
There are 203 file descriptors available, 163 required, 203 reserved, and 288 requested.
Warning: Reducing -maxconnections from 125 to 40, because of file descriptor limitations.
...
Using at most 40 automatic connections (203 file descriptors available)
~~~

**Coverage with this patch + 1024 ulimit (matches most Linux systems)**
~~~
Bitcoin Core version v0.18.99.0-50ccaa56f (release build)
There are 1024 file descriptors available, 163 required, 288 reserved, and 288 requested.
...
Using at most 125 automatic connections (1024 file descriptors available)
~~~

The replaced (old) arithmetic smells strange if you consider the subtraction can subtract negative values to increase their value, and any unintended behavior will then be silenced by std::max(x,0). Now there should be less likelyhood of suppressed failure.

Possible fix for #14870

Edit: thanks to IRC chat for the tips!

### CODE REVIEW

**Review by jonatack — COMMENTED** [review]
Reference IRC discussion, IIUC: http://www.erisian.com.au/bitcoin-core-dev/log-2019-05-10.html#l-127

Tests that show the expected behavior and verify the issue/regression would be good here if possible.

File: src/net.h [hunk 1]
```diff
@@ -59,6 +59,8 @@ static const unsigned int MAX_SUBVERSION_LENGTH = 256;
 static const int MAX_OUTBOUND_CONNECTIONS = 8;
 /** Maximum number of addnode outgoing nodes */
 static const int MAX_ADDNODE_CONNECTIONS = 8;
+/** Number of feeler connections */
+static const int NUM_FEELER_CONNECTIONS = 1;
```
Comment (Empact) [inline src/net.h]: nit: `MAX_FEELER_CONNECTIONS` for consistency, and coherence with `nMaxFeeler`

**Comment by Empact** [comment]
Two requests to simplify review:
* squash travis fixups into the appropriate commit
* split fixes into 2 PRs, one addressing each

**Comment by tryphe** [comment]
@Empact Thanks. I thought it would be crazy to open two PRs, because bug 1 fixes something for sure, and bug 2 isn't really reproducible. Any bad behavior from bug 2 could be confused for bug 1, and it would be hard to tell which one it is. Eg, things could go south if not enough FDs were allocated _or_ not enough were asserted. And since file descriptor problems via threads are hard to reproduce, I thought it made sense to mash it all together.

File: src/net.h (hunk 1 above)
Comment (tryphe) [reply src/net.h]: Right, that makes sense. This essentially becomes 0 when maxconnections < MAX_ADDNODE_CONNECTIONS, which I wasn't sure of initially.

**Review by tryphe — COMMENTED** [review]
Looking for more suggestions here. I'm not quite sure how to logically split this into multiple PRs, so I'll leave that for now.

This commit certainly isn't perfect, but is essentially an open improvement to enforce/allocate a slightly more accurate number of file descriptors. If the user has an absurdly low system setting, we lower the number of maximum inbound connections until it's no longer possible, instead of possibly passing with an unsafe amount. 

One note: I've included mmaped things as well, because they require a temporary handle to obtain the mapping. Not sure if this was overkill but it seemed like a good idea.

I also have a feeling that the number of bind arguments isn't always equal to the number of interfaces bitcoind will listen on, so that's probably a bad assumption. 

Any additions or ideas on how to fix this up is appreciated. Thanks :)

File: src/init.cpp [hunk 2]
```diff
@@ -1010,12 +1010,12 @@ bool AppInitParameterInteraction()
...
 
     // Calculate new -maxconnection count. Note: std::min<int>(...) to work around FreeBSD compilation issue described in #2695
-    nMaxConnections = std::max(std::min<int>(nMaxConnections, nFD - nFDMin), 0);
+    nMaxConnections = std::min<int>(nMaxConnections, nFD - nFDMin);
     LogPrintf("There are %d file descriptors available, %d required, %d reserved, and %d requested.\n", nFD, nFDMin, nFDMin + nMaxConnections, nFDMin + nUserMaxConnections);
 
     // Display a warning if the number of total maximum connections was lowered
     if (nMaxConnections < nUserMaxConnections)
-        InitWarning(strprintf(_("Reducing -maxconnections from %d to %d, because of file descriptor limitations."), nUserMaxConnections, nMaxConnections));
+        InitWarning(strprintf(_("Reducing -maxconnections from %d to %d, because of system limitations."), nUserMaxConnections, nMaxConnections));
```
Comment (tryphe) [inline src/init.cpp]: nit: maybe I should remove this change for log compatability

**Comment by tryphe** [comment]
Maybe we can just implement `-extrahandles=n' instead of modifying this code every time we realize something new or existing needs a file handle. It doesn't really make sense to modify this code if we can just modify a default config value, no?

**Comment by tryphe** [comment]
I figured I should supply some pseudocode to show the result what happens differently in the code and how it's related to both Bug 1 and 2:

Suppose you are running Mac OS X and listening on the network interface. The ulimit is 256. Suppose you also have 1 bind interface `nBind(1)` and 1 RPC thread `nRpc(1)`.

**Current code**
~~~
nFD = RaiseFileDescriptorLimit(nMaxConn(125) + nCore(150) + nAddnode(8)) = gets reduced to 256
nMaxConnections = min(125, nFD(256) - nBind(1) - nCore(150) - nAddnode(8) = min(125, 97) = 97
if ( nFD < nCore(150) ) error("not enough fds")
nMaxConnections = min(nFD(256) - nCore(150) - nAddnode(8), nMaxConnections(97)) = min(98, 97) = 97
~~~

Usable FDs: nMinCoreFDs(150) + nMaxConnections(97) = 247 + nAddnode(8) + nRpc(1) + nBind(1) = 257
There's too many, even if we only have 1 bind interface and 1 RPC thread.


**This PR**
~~~
nFDMin = nCore(150) + nAddnode(8) + nBind(1) + nRpc(1) = 160
nFD = RaiseFileDescriptorLimit(nMaxConn(125) + nFDMin(160)) = gets reduced to 256
if ( nFD < nFDMin(160) ) error("not enough fds")
nMaxConnections = min(nMaxConn(125), 256-nFDMin(160)) = 96
~~~

FDs accounted for: nCore(150) + nMaxConnections(96) = 246 + nAddnode(8) + nRpc(1) + nBind(1) = 256

**Comment by adamjonas** [comment]
Hi @tryphe - Wondering if you are going to continue with this PR or whether it should be closed. It looks like this has needed a rebase for some time. I'd also suggest you squash your commits in accordance with the [CONTRIBUTING](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits) guidelines and add tests as [suggested above](https://github.com/bitcoin/bitcoin/pull/16003#pullrequestreview-237742966) to verify the issue.

**Comment by fanquake** [comment]
Will close this for now. However I think the changes are worth following up on, so will open a new good first issue. Will also extract one change that can just be merged now.

**Comment by tryphe** [comment]
Thanks for following up @adamjonas @fanquake!

I certainly don't think this is the most correct way to proceed, especially with the large scope of changes with new sockets and threads. But it's closer than what we had before, in terms of knowing the [required, requested, allocated] counts of FDs. I think it makes much more sense to modularize the summing of FD contexts and make it easier to count the required amount of threads with future changes. Maybe something like a `DescriptorMan` that does all of the dirty work. Doing this will make future changes like more socket or modifying connection min/max bounds much less conflicted. It's apparent that we don't want to keep editing these general files (init.cpp and net.h) every time we need to change the descriptor count. 

Will ping #18911 so they can see this message.

**Pull request closed without merging**